### PR TITLE
Bugfix/ota 572/handle manifest only once

### PR DESCRIPTION
**Pull request by jochenschneider**
Created an "incoming cache" for manifests to handle the same manifest only once. It needs to be handled again when the targets change, though, since manifest error handling depends on the current target (the manifest and target.json are expected to contain the same packages).

### CODE REVIEW

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 1]
```diff
@@ -1,28 +1,47 @@
...
                            verifier: TufKey => Verifier
                           )(implicit val db: Database, val ec: ExecutionContext)
-    extends DeviceRepositorySupport
-    with UpdateTypesRepositorySupport {
+    extends DeviceRepositorySupport with UpdateTypesRepositorySupport with ProcessedManifestsRepositorySupport {
   private lazy val _log = LoggerFactory.getLogger(this.getClass)
+  val digester: MessageDigest = MessageDigest.getInstance("SHA-256")
+
+  def setDeviceManifest(namespace: Namespace, device: DeviceId, signedDevMan: SignedPayload[Json]): Future[Unit] = {
+    val digest = digester.digest(signedDevMan.signed.canonical.getBytes("UTF-8"))
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: You can use `com.advancedtelematic.libtuf_server.crypto.Sha256Digest.digest().hash.get` 

But this is a bit finicky, if a space gets added to the manifest or something then it will be processed again

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 2]
```diff
@@ -1,28 +1,47 @@
...
+    val hex = digest.map("%02x".format(_)).mkString
+
+    processedManifestsRepository.contains(namespace, device, hex).flatMap { alreadyProcessed =>
+      if (alreadyProcessed) {
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: Could be:

```scala
    processedManifestsRepository.contains(namespace, device, hex).flatMap {
      case true =>
        FastFuture.successful(())
      case false =>
        for {
          _ <- processDeviceManifest(namespace, device, signedDevMan)
          _ <- processedManifestsRepository.add(namespace, device, hex)
        } yield ()
```

But I guess that's just style.

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 3]
```diff
@@ -1,28 +1,47 @@
...
+        FastFuture.successful(())
+      } else {
+        processDeviceManifest(namespace, device, signedDevMan).flatMap { _ =>
+          processedManifestsRepository.add(namespace, device, hex).map(_ => ())
+        }
+      }
+    }
+  }
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: Could we somehow incorporate this check with the current states? Because currently there is already `NoChange` [1], isn't this an extension of that? 

[1]: https://github.com/advancedtelematic/director/blob/53de30fb27bfefd4c7a0c9bfeb0ab35193ff3a01/src/main/scala/com/advancedtelematic/director/manifest/AfterUpdate.scala#L44

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala (hunk 1 above)
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: I think the chance that the manifest format changes are low and it also won't break things.

Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: OTOH what could happen is that `timeserver_time`s get corruptly filled in and change all of a sudden?

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala (hunk 3 above)
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: That's a good idea I didn't consider, but my mental model is more that of an "incoming cache". I find the code in `DeviceUpdate` already hard to follow and would think adding another `DeviceUpdateResult` would make it even worse.

Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: The code in `DeviceManifestUpdate` is simpler I guess.

**Comment by jochenschneider** [comment]
Updated with small changes.

File: src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala [hunk 4]
```diff
@@ -388,10 +400,18 @@ trait DeviceResourceSpec extends DirectorSpec with KeyGenerators with DefaultPat
 
     updateManifestOk(device, deviceManifest)
 
+    val digest = digester.digest(deviceManifest.signed.canonical.getBytes("UTF-8"))
+    val hex = digest.map("%02x".format(_)).mkString
```
Comment (simao) [inline src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala]: Can also use `Sha256Digest.digest` here

File: src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala [hunk 5]
```diff
@@ -454,8 +474,8 @@ trait DeviceResourceSpec extends DirectorSpec with KeyGenerators with DefaultPat
     // currently device-current-target and device-update-target are both at 2
     // sending empty device manifest should not update device-current-target to 3
     updateManifestOk(device, deviceManifest2)
-    deviceVersion(device) shouldBe Some(3)
-    deviceScheduledVersion(device) shouldBe 3
+    deviceVersion(device) shouldBe Some(2)
```
Comment (simao) [inline src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala]: this should be written: `should contain 2` so that the matcher gives a nicer msg if it fails

File: src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala [hunk 6]
```diff
@@ -336,6 +339,13 @@ trait DeviceResourceSpec extends DirectorSpec with KeyGenerators with DefaultPat
     updateManifestOk(device, deviceManifestTarget)
 
     coreClient.getReport(updateId) shouldBe Seq(operation)
+    fetchTargetsFor(device).signed.targets.isEmpty shouldBe true
+
+    // schedule and confirm again
+    schedule(device, targets, updateId)
+    fetchTargetsFor(device).signed.targets.isEmpty shouldBe false
```
Comment (simao) [inline src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala]: this should be `fetchTargetsFor(device).signed.targets shouldBe empty` for a better message on failure

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala [hunk 7]
```diff
@@ -1,28 +1,50 @@
...
   private lazy val _log = LoggerFactory.getLogger(this.getClass)
+  val digester: MessageDigest = MessageDigest.getInstance("SHA-256")
+
+  def setDeviceManifest(namespace: Namespace, device: DeviceId, signedDevMan: SignedPayload[Json]): Future[Unit] = {
+    val signed = signedDevMan.signed
+    val timeless = signed.mapObject(_.remove("timeserver_time").remove("previous_timeserver_time"))
+    val digest = Sha256Digest.digest(timeless.canonical.getBytes("UTF-8")).hash.value
 
-  def setDeviceManifest(namespace: Namespace, device: DeviceId, signedDevMan: SignedPayload[Json]): Future[Unit] = for {
+    processedManifestsRepository.contains(namespace, device, digest).flatMap { alreadyProcessed =>
```
Comment (simao) [inline src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: you didn't like the case/for ?

File: src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala (hunk 5 above)
Comment (jochenschneider) [reply src/test/scala/com/advancedtelematic/director/http/DeviceResourceSpec.scala]: Ah cool!

File: src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala (hunk 7 above)
Comment (jochenschneider) [reply src/main/scala/com/advancedtelematic/director/manifest/DeviceManifestUpdate.scala]: I don't like mixing flatMaps and fors, I'll change the if/else to case true/false.

**Comment by jochenschneider** [comment]
Updated.
